README.md
[![Build Status](https://travis-ci.org/chef-partners/dcos-cookbook.svg?branch=master)](https://travis-ci.org/chef-partners/dcos-cookbook)

Description
===========

Manage deployment and configuration of underlying Mesosphere DC/OS installation. This cookbook supports both DC/OS
OSS installations (default) and DC/OS Enterprise.

Note: The default username and password for DC/OS Enterprise is `bootstrapuser`/`deleteme` and can be changed by setting
`node['dcos']['config']['superuser_username']` and `node['dcos']['config']['superuser_password_hash']`, respectively.

Requirements
------------

Only Red Hat or CentOS 7.x are currently supported.

Usage
==========

The behavior of this cookbook is managed by attributes documented in the [attributes file](attributes/default.rb).
The `node['dcos']['dcos_role']` attribute controls the DC/OS role to apply to the node (default is `master`). The
`node['dcos']['config']['master_list']` must be set when `node['dcos']['config']['master_discovery']` is set to
`static` to specify the list of DC/OS master node IPv4 addresses to connect at startup (this must be an odd number
of masters and cannot be changed, later).

This cookbook uses the stable channel, by default. Setting `node['dcos']['dcos_version']` to `earlyaccess` will
install the latest EarlyAccess version of DC/OS OSS. Enterprise support can be enabled by setting
`node['dcos']['dcos_enterprise']` to `true` and providing a license key in `node['dcos']['dcos_license_text']`
via whichever manner you prefer.

Roles
----------

You can create a Chef Role and apply it to nodes as necessary to specify `master`, `slave` and `slave_public`, as
appropriate. Any additional configuration should be set as override attributes in an Environment to ensure that
all nodes receive those global settings.

### Example Role dcos_master.rb ###
````ruby
name "dcos_master"
description "DC/OS master role"
run_list "recipe[dcos]"
default_attributes "dcos" => {
    "dcos_role" => "master",
    "config" => {
        "master_list" => [ "10.0.2.10" ]
    }
}
````

### Example Role dcos_slave.rb ###
````ruby
name "dcos_slave"
description "DC/OS slave role"
run_list "recipe[dcos]"
default_attributes "dcos" => {
    "dcos_role" => "slave",
    "config" => {
        "master_list" => [ "10.0.2.10" ]
    }
}
````

Recipe
=======

default
-------

Installs the prerequisites for the Mesosphere DC/OS installation, including packages, and groups.
Docker with OverlayFS is installed and enabled if `node['dcos']['manage_docker']` is set to true,
which is the default. Next, the recipe downloads and runs the installation package with the
settings configured by the attributes under `node['dcos']['config']`.


Resource
========

dcos_user
---------

Defines a DC/OS user.

### Example dcos_user ###
````ruby
dcos_user 'user@domain.com' do
  email 'user@domain.com'
end
````

It's possible Zookeeper is not up and running at the time you attempt to
provision a DC/OS user. If that is the case, you should add an
`ignore_failure` set to `true` in the used declaration:

### Example dcos_user with ignore_failure ###
````ruby
dcos_user 'user@domain.com' do
  ignore_failure true
  email 'user@domain.com'
end
````

Testing
=======

ChefSpec
--------
There is basic coverage for the default recipe.

InSpec
------
There is basic functional testing of the DC/OS OSS setup for a single master node.

Test Kitchen
------------
The included [.kitchen.yml](.kitchen.yml) runs the default master deployment in a generic fashion.
The included [.kitchen.local.yml.example](.kitchen.local.yml.example) shows alternate settings for
running multi-master with slaves on GCE (you will have to rename and update accordingly).

License and Author
==================

Author:: Matt Ray (<matt@chef.io>)

Copyright 2016-2018 Chef Software, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log

## [v3.0.1](https://github.com/chef-partners/dcos-cookbook/tree/v3.0.1)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v3.0.0...v3.0.1)

**Merged pull requests:**

- Simplify download URL generation [\#31](https://github.com/chef-partners/dcos-cookbook/pull/31) ([wolf31o2](https://github.com/wolf31o2))

## [v3.0.0](https://github.com/chef-partners/dcos-cookbook/tree/v3.0.0)

**BREAKING CHANGES**

This cookbook has dropped support for DC/OS versions prior to 1.10.0, which are no longer supported by Mesosphere.

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v2.2.0...v3.0.0)

**Merged pull requests:**

- Drop support for DC/OS older than 1.10 [\#29](https://github.com/chef-partners/dcos-cookbook/pull/29) ([wolf31o2](https://github.com/wolf31o2))
- Support DC/OS 1.12.2, 1.11.10, and 1.10.11 [\#30](https://github.com/chef-partners/dcos-cookbook/pull/30) ([wolf31o2](https://github.com/wolf31o2))

## [v2.2.0](https://github.com/chef-partners/dcos-cookbook/tree/v2.2.0)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v2.1.0...v2.2.0)

**Merged pull requests:**

- Support DC/OS 1.12.1, 1.11.8, and 1.11.9 [\#28](https://github.com/chef-partners/dcos-cookbook/pull/28) ([wolf31o2](https://github.com/wolf31o2))

## [v2.1.0](https://github.com/chef-partners/dcos-cookbook/tree/v2.1.0)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v2.0.2...v2.1.0)

**Merged pull requests:**

- Support DC/OS 1.12.0 and 1.11.7 [\#27](https://github.com/chef-partners/dcos-cookbook/pull/27) ([wolf31o2](https://github.com/wolf31o2))

## [v2.0.2](https://github.com/chef-partners/dcos-cookbook/tree/v2.0.2)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v2.0.1...v2.0.2)

**Merged pull requests:**

- Support DC/OS 1.11.5 and 1.11.6 [\#26](https://github.com/chef-partners/dcos-cookbook/pull/26) ([wolf31o2](https://github.com/wolf31o2))

## [v2.0.1](https://github.com/chef-partners/dcos-cookbook/tree/v2.0.1)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v2.0.0...v2.0.1)

**Merged pull requests:**

- Support DC/OS 1.11.4, and latest 1.9 and 1.10 [\#25](https://github.com/chef-partners/dcos-cookbook/pull/25) ([wolf31o2](https://github.com/wolf31o2))

## [v2.0.0](https://github.com/chef-partners/dcos-cookbook/tree/v2.0.0)

**BREAKING CHANGES**

This cookbook has upgraded its dependencies and now requires at least Chef 12.15+

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.3.2...v2.0.0)

**Merged pull requests:**

- Use latest docker cookbook [\#24](https://github.com/chef-partners/dcos-cookbook/pull/24) ([wolf31o2](https://github.com/wolf31o2))

## [v1.3.2](https://github.com/chef-partners/dcos-cookbook/tree/v1.3.2)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.3.1...v1.3.2)

**Merged pull requests:**

- Support DC/OS 1.11.3 [\#23](https://github.com/chef-partners/dcos-cookbook/pull/23) ([wolf31o2](https://github.com/wolf31o2))

## [v1.3.1](https://github.com/chef-partners/dcos-cookbook/tree/v1.3.1)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.3.0...v1.3.1)

**Merged pull requests:**

- Support DC/OS 1.11.2 [\#22](https://github.com/chef-partners/dcos-cookbook/pull/22) ([wolf31o2](https://github.com/wolf31o2))

## [v1.3.0](https://github.com/chef-partners/dcos-cookbook/tree/v1.3.0)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.2.3...v1.3.0)

**Merged pull requests:**

- Recipe for DC/OS CLI [\#17](https://github.com/chef-partners/dcos-cookbook/pull/17) ([wolf31o2](https://github.com/wolf31o2))
- Updates from session w/ Justin Lee [\#20](https://github.com/chef-partners/dcos-cookbook/pull/20) ([wolf31o2](https://github.com/wolf31o2)) ([justinrlee](https://github.com/justinrlee))
- Allow disabling scripts using nil source [\#21](https://github.com/chef-partners/dcos-cookbook/pull/21) ([wolf31o2](https://github.com/wolf31o2)) 

## [v1.2.3](https://github.com/chef-partners/dcos-cookbook/tree/v1.2.3)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.2.2...v1.2.3)

**Merged pull requests:**

- Guard cloud fault-domain-detect script when onprem [\#16](https://github.com/chef-partners/dcos-cookbook/pull/16) ([wolf31o2](https://github.com/wolf31o2))
- Helper: dcos_cluster_address [\#18](https://github.com/chef-partners/dcos-cookbook/pull/18) ([wolf31o2](https://github.com/wolf31o2))
- Support DC/OS 1.10.6 and 1.11.1 [\#19](https://github.com/chef-partners/dcos-cookbook/pull/19) ([wolf31o2](https://github.com/wolf31o2))

## [v1.2.2](https://github.com/chef-partners/dcos-cookbook/tree/v1.2.2)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.2.1...v1.2.2)

**Merged pull requests:**

- Create a preflight-check execute block [\#15](https://github.com/chef-partners/dcos-cookbook/pull/15) ([wolf31o2](https://github.com/wolf31o2))

## [v1.2.1](https://github.com/chef-partners/dcos-cookbook/tree/v1.2.1)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.2.0...v1.2.1)

**Merged pull requests:**

- Remove zookeeper Gem requirements [\#14](https://github.com/chef-partners/dcos-cookbook/pull/14) ([wolf31o2](https://github.com/wolf31o2))

## [v1.2.0](https://github.com/chef-partners/dcos-cookbook/tree/v1.2.0)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.1.0...v1.2.0)

**Merged pull requests:**

- Add dcos_user custom resource [\#10](https://github.com/chef-partners/dcos-cookbook/pull/10) ([rbanffy](https://github.com/rbanffy))
- Support DC/OS Enterprise [\#13](https://github.com/chef-partners/dcos-cookbook/pull/13) ([wolf31o2](https://github.com/wolf31o2))

## [v1.1.0](https://github.com/chef-partners/dcos-cookbook/tree/v1.1.0)

[Full Changelog](https://github.com/chef-partners/dcos-cookbook/compare/v1.0.0...v1.1.0)

**Merged pull requests:**


spec/unit/recipes/cli_spec.rb
#
# Cookbook Name:: dcos
# Spec:: cli
#

require 'spec_helper'

describe 'dcos::cli' do
  context 'Default behavior' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: '7.3.1611').converge(described_recipe)
    end

    it 'creates remote_file[/usr/local/bin/dcos]' do
      expect(chef_run).to create_remote_file('/usr/local/bin/dcos').with(mode: '0755')
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: dcos
# Spec:: default
#

require 'spec_helper'

describe 'dcos::default' do
  context 'Default behavior, assume eth0' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: '7.3.1611') do |node|
        node.override['network']['interfaces']['eth0']['addresses'] =
          {
            '1.2.3.4' =>
              {
                'family' => 'inet',
                'netmask' => '255.255.255.0',
                'broadcast' => '192.168.1.255',
              },
          }
        stub_command(/ping -c1 /).and_return(false)
      end.converge(described_recipe)
    end

    it 'sets SELinux to permissive' do
      expect(chef_run).to permissive_selinux_state('SELinux Permissive')
    end

    %w(
      firewalld
      rsyslog
    ).each do |svc|
      it "stops and disables #{svc}" do
        expect(chef_run).to stop_service(svc)
        expect(chef_run).to disable_service(svc)
      end
    end

    it 'installs %w(curl ipset tar unzip xz)' do
      expect(chef_run).to install_package(%w(curl ipset tar unzip xz))
    end

    it 'installs net-tools' do
      expect(chef_run).to install_package('net-tools')
    end

    it 'creates group[nogroup]' do
      expect(chef_run).to create_group('nogroup')
    end

    it 'creates docker service' do
      expect(chef_run).to create_docker_service('default')
    end

    it 'creates remote_file[dcos_generate_config.sh]' do
      expect(chef_run).to create_remote_file('/usr/src/dcos/dcos_generate_config.sh').with(mode: '0755')
    end

    it 'creates directory[genconf]' do
      expect(chef_run).to create_directory('/usr/src/dcos/genconf').with(mode: '0755')
    end

    it 'creates template[config.yaml]' do
      expect(chef_run).to create_template('/usr/src/dcos/genconf/config.yaml')
    end

    it 'creates /usr/src/dcos/genconf/ip-detect from template' do
      expect(chef_run).to create_template('/usr/src/dcos/genconf/ip-detect')
    end

    it 'creates /usr/src/dcos/genconf/ip-detect-public from file' do
      expect(chef_run).to create_cookbook_file('/usr/src/dcos/genconf/ip-detect-public')
    end

    it 'executes[dcos-genconf]' do
      expect(chef_run).to run_execute('dcos-genconf').with(user: 'root')
    end

    it 'ensures /usr/src/dcos/genconf/serve/dcos_install.sh is executable' do
      expect(chef_run).to create_file('/usr/src/dcos/genconf/serve/dcos_install.sh').with(mode: '0755')
    end

    it 'executes[preflight-check]' do
      expect(chef_run).to run_execute('preflight-check')
    end

    it 'executes[dcos_install]' do
      expect(chef_run).to run_execute('dcos_install').with(user: 'root')
    end

    %w(
      mesos
      marathon
    ).each do |svc|
      it "executes[check-#{svc}-up]" do
        expect(chef_run).to run_execute("check-#{svc}-up")
      end
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'DC/OS Enterprise 1.11' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'centos', version: '7.3.1611') do |node|
        node.override['dcos']['dcos_enterprise'] = true
        node.override['dcos']['dcos_version'] = '1.11.1'
        node.override['dcos']['dcos_license_text'] = 'STUB'
        stub_command(/ping -c1 /).and_return(false)
      end.converge(described_recipe)
    end

    it 'creates genconf/license.txt' do
      expect(chef_run).to create_file('/usr/src/dcos/genconf/license.txt')
    end

    it 'creates genconf/fault-domain-detect' do
      expect(chef_run).to create_cookbook_file('/usr/src/dcos/genconf/fault-domain-detect')
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/controls/default_spec.rb
#
# Copyright:: (c) 2017 Chris Gianelloni
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

%w(
  firewalld
  rsyslog
).each do |svc|
  describe service(svc) do
    it { should_not be_running }
    it { should_not be_enabled }
  end
end

%w(
  curl
  ipset
  tar
  unzip
  xz
).each do |pkg|
  describe package(pkg) do
    it { should be_installed }
  end
end

describe group('nobody') do
  it { should exist }
end

describe directory('/opt/mesosphere') do
  it { should exist }
  it { should be_directory }
  it { should be_readable }
end

describe file('/opt/mesosphere/environment') do
  it { should exist }
  it { should be_file }
  it { should be_readable }
end

describe file('/etc/profile.d/dcos.sh') do
  it { should exist }
  it { should be_file }
  it { should be_readable }
  it { should be_symlink }
end

target_file = file('/etc/profile.d/dcos.sh').link_path
describe file(target_file) do
  it { should exist }
  it { should be_file }
  it { should be_readable }
end

%w(
  docker
  dcos-exhibitor
  dcos-marathon
  dcos-mesos-master
).each do |svc|
  describe service(svc) do
    it { should be_installed }
    it { should be_enabled }
    it { should be_running }
  end
end

describe file('/etc/mesosphere/roles/master') do
  it { should exist }
  it { should be_readable }
end

%w(
  slave
  slave_public
).each do |role|
  describe file("/etc/mesosphere/roles/#{role}") do
    it { should_not exist }
  end
end

%w(
  /etc/mesosphere/setup-flags/repository-url
).each do |conf|
  describe file(conf) do
    it { should exist }
    it { should be_readable }
  end
end

describe yaml('/etc/rexray/config.yml') do
  its(['rexray', 'modules', 'default-docker', 'disabled']) { should eq true }
end

# IPv4 (TCP + UDP)
describe port(53) do
  it { should be_listening }
  its('protocols') { should include 'tcp' }
  its('protocols') { should include 'udp' }
end

# IPv4 (TCP + TCP6)
%w(
  80
  443
  2181
  5050
  8080
  8181
).each do |tcp|
  describe port(tcp) do
    it { should be_listening }
    its('protocols') { should cmp(/tcp/) }
  end
end


recipes/cli.rb
#
# Cookbook Name:: dcos
# Recipe:: cli
#
# Copyright 2018 Chris Gianelloni
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

cli_base_url = "https://downloads.dcos.io/binaries/cli/#{node['os']}/#{node['kernel']['machine'].tr('_', '-')}"

# Fetch the `dcos` CLI tool
remote_file '/usr/local/bin/dcos' do
  source "#{cli_base_url}/dcos-#{node['dcos']['dcos_version'].to_f}/dcos"
  mode '0755'
end


libraries/helpers.rb
module Dcos
  # Helper functions
  module Helpers
    def dcos_generate_config_url
      return node['dcos']['dcos_generate_config_url'] if node['dcos'].key?('dcos_generate_config_url')
      return "#{dcos_base_url}/dcos_generate_config.ee.sh" if dcos_enterprise?
      "#{dcos_base_url}/dcos_generate_config.sh"
    end

    def dcos_enterprise?
      node['dcos']['dcos_enterprise'].to_s == 'true'
    end

    def dcos_cluster_address
      if node['dcos'].key?('config') && node['dcos']['config'].key?('master_discovery') &&
         node['dcos']['config']['master_discovery'] == 'master_http_loadbalancer' &&
         node['dcos']['config'].key?('master_external_loadbalancer')
        node['dcos']['config']['master_external_loadbalancer']
      elsif node['dcos'].key?('config') && node['dcos']['config'].key?('master_list')
        node['dcos']['config']['master_list'].first # yuck, we can do better
      else
        'leader.mesos' # assuming we're inside the cluster
      end
    end

    private

    def dcos_base_url
      v = node['dcos']['dcos_version']
      return 'https://downloads.dcos.io/dcos/EarlyAccess' if v.downcase == 'earlyaccess'
      if v.to_f >= 1.10
        return "https://downloads.mesosphere.com/dcos-enterprise/stable/#{v}" if dcos_enterprise?
        return "https://downloads.dcos.io/dcos/stable/#{v}"
      else # stable or older releases
        return 'https://downloads.mesosphere.com/dcos-enterprise/stable/1.12.2' if dcos_enterprise?
        'https://downloads.dcos.io/dcos/stable'
      end
    end
  end
end

Chef::Recipe.send(:include, Dcos::Helpers)
Chef::Resource.send(:include, Dcos::Helpers)


recipes/_ip-detect.rb
#
# Cookbook Name:: dcos
# Recipe:: _ip-detect
#
# Copyright 2016, Chef Software, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# generates a genconf/ip-detect script based on the node['dcos']['ip-detect']
# https://docs.mesosphere.com/archived-dcos-enterprise-edition/installing-enterprise-edition-1-5/create-a-script-for-ip-address-discovery/

if %w(aws gce).include?(node['dcos']['ip-detect'])
  cookbook_file '/usr/src/dcos/genconf/ip-detect' do
    source node['dcos']['ip-detect']
    mode '0755'
  end
  cookbook_file '/usr/src/dcos/genconf/ip-detect-public' do
    source "#{node['dcos']['ip-detect']}_public"
    mode '0755'
  end
else
  # find the ipaddress for that interface
  interface = node['dcos']['ip-detect']
  template '/usr/src/dcos/genconf/ip-detect' do
    source 'ip-detect.erb'
    mode '0755'
    variables(interface: interface)
  end
  cookbook_file '/usr/src/dcos/genconf/ip-detect-public' do
    cookbook node['dcos']['ip-detect-public']['cookbook']
    source node['dcos']['ip-detect-public']['source']
    mode '0755'
  end
end


resources/dcos_cli.rb
#
# Cookbook Name:: dcos
# Resource:: dcos_cli
#
# Copyright 2018, Chris Gianelloni
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

resource_name :dcos_cli

# standard attributes for this resource
property :command, String, name_property: true
property :cluster_uri, String, required: false, default: 'http://leader.mesos'
property :cli_path, String, default '/usr/local/bin/dcos'
property :ssl_verify, [true, false], default: false

# copied from execute resource
property :cwd, String
property :environment, Hash, default: {}
property :user, [String, Integer]
property :sensitive, [true, false], default: false

# optional authentication
property :username, String, required: false, default: nil
property :password, String, sensitive: true, required: false, default: nil

action :run do
  # set some empty variables
  run_env = {}
  user_pass = ''
  ssl_verify = ''

  # set some other variables
  cli = new_resource.cli_path

  if new_resource.username && new_resource.password
    run_env = new_resource.environment.merge({'DCOS_PASSWORD': new_resource.password})
    user_pass = "--username=#{new_resource.username} --password-env=DCOS_PASSWORD"
  end

  ### TODO: make this support older DC/OS versions, OSS versions, etc.
  execute 'dcos-cluster-setup' do
    command "#{cli} cluster setup #{new_resource.cluster_uri} #{ssl_verify} #{user_pass}"
    environment run_env
    sensitive true
    only_if "#{cli} cluster --info"
  end

  # assume that we're authenticated, or don't need authentication
  execute 'dcos-cli' do
    command "#{cli} #{new_resource.command}"
    cwd new_resource.cwd
    environment new_resource.environment
    sensitive new_resource.sensitive
    user new_resource.user
  end
end


resources/dcos_user.rb
#
# Cookbook Name:: dcos
# Resource:: dcos_user
#
# Copyright 2018, Chef Software, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

resource_name :dcos_user

property :zk_host, String,
         default: 'zk-1.zk:2181,zk-2.zk:2181,zk-3.zk:2181,zk-4.zk:2181,'\
                  'zk-5.zk:2181',
         required: true
property :email, String, required: false

load_current_value do
  require 'zookeeper'
  include Zookeeper::Constants
  z = Zookeeper.new(zk_host)
  user_node = z.get(path: "/dcos/users/#{email}")
  email user_node[:data] if user_node[:rc] == ZOK
end

action :create do
  # If there is a change, remove and replace the current data
  converge_if_changed :email do
    require 'zookeeper'
    include Zookeeper::Constants
    z = Zookeeper.new(zk_host)
    z.delete(path: "/dcos/users/#{email}") # Fails cleanly if it doesn't exist.
    z.create(path: "/dcos/users/#{email}", data: email)
  end
end

action :delete do
  require 'zookeeper'
  include Zookeeper::Constants
  # Remove the user node from Zookeeper
  z = Zookeeper.new(zk_host)
  z.delete(path: "/dcos/users/#{email}")
end


recipes/default.rb
#
# Cookbook Name:: dcos
# Recipe:: default
#
# Copyright 2016, Chef Software, Inc.
# Copyright 2017-2018 Chris Gianelloni
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Prereqs
include_recipe 'selinux::permissive'

%w(
  firewalld
  rsyslog
).each do |svc|
  service svc do
    action %i(stop disable)
  end
end

package %w(
  curl
  ipset
  tar
  unzip
  xz
)

# Hard-coded `ifconfig` means we have to include this
# https://github.com/apache/mesos/blob/a741b15e889de3242e3aa7878105ab9d946f6ea2/src/slave/containerizer/mesos/isolators/network/cni/cni.cpp#L2106
package 'net-tools'

group 'nogroup'

# Install docker with overlayfs
docker_service 'default' do
  storage_driver node['dcos']['docker_storage_driver']
  version node['dcos']['docker_version'] if node['dcos']['docker_version']
  install_method 'package' if node['dcos']['docker_version']
  action %i(create start)
  only_if { node['dcos']['manage_docker'] }
end

directory '/usr/src/dcos/genconf' do
  mode '0755'
  recursive true
end

# Only used on DC/OS Enterprise 1.11+
file '/usr/src/dcos/genconf/license.txt' do
  content node['dcos']['dcos_license_text']
  sensitive true
  only_if { dcos_enterprise? && node['dcos']['dcos_version'].to_f >= 1.11 }
end

template '/usr/src/dcos/genconf/config.yaml' do
  source 'config.yaml.erb'
  variables config: node['dcos']['config']
end

# Only supported on DC/OS Enterprise 1.11+
cookbook_file '/usr/src/dcos/genconf/fault-domain-detect' do
  # Pull latest from GitHub
  cookbook node['dcos']['fault-domain-detect']['cookbook']
  source node['dcos']['fault-domain-detect']['source']
  mode '0755'
  only_if { dcos_enterprise? && node['dcos']['dcos_version'].to_f >= 1.11 }
  not_if { node['dcos']['fault-domain-detect']['source'].nil? }
end

remote_file '/usr/src/dcos/dcos_generate_config.sh' do
  source dcos_generate_config_url
  mode '0755'
end

# generate the genconf/ip-detect script
include_recipe 'dcos::_ip-detect'

execute 'dcos-genconf' do
  command '/usr/src/dcos/dcos_generate_config.sh --genconf'
  user 'root'
  cwd '/usr/src/dcos'
  creates '/usr/src/dcos/genconf/cluster_packages.json'
end

file '/usr/src/dcos/genconf/serve/dcos_install.sh' do
  mode '0755'
end

# We're going to poll this up to 30 times, to prevent issues where a port may temporarily be in use
execute 'preflight-check' do
  command "/usr/src/dcos/genconf/serve/dcos_install.sh --preflight-only #{node['dcos']['dcos_role']}"
  retry_delay 1
  retries 30
  action :run
  not_if { ::File.exist?('/opt/mesosphere/environment') } # assume DC/OS is installed
end

execute 'dcos_install' do
  command "/usr/src/dcos/genconf/serve/dcos_install.sh #{node['dcos']['dcos_role']}"
  creates '/opt/mesosphere/environment'
  user 'root'
  cwd '/usr/src/dcos'
  action :run
end

# Poll for leader.mesos to ensure we're up before finishing converge
execute 'check-mesos-up' do
  command 'ping -c1 leader.mesos'
  retry_delay 1
  retries node['dcos']['leader_check_retries']
  action :run
  not_if 'ping -c1 leader.mesos'
end

execute 'check-marathon-up' do
  command 'ping -c1 marathon.mesos'
  retry_delay 1
  retries node['dcos']['leader_check_retries']
  action :run
  not_if 'ping -c1 marathon.mesos'
end
